### ISSUE TITLE
Enable DEX authz in Sidecar Proxy Instead of Ingress Gateway

### ISSUE THREAD

**Issue by aaron-arellano** [kubeflow/manifests#1834]
Hello Manifests. I am re-opening an issue based on this one from last year: [https://github.com/kubeflow/manifests/issues/1191]. We are still exploring how to configure the EnvoyFilter for DEX authentication to authenticate on kubeflow service sidecars instead of having the filter at the ingress gateway. 

Like the previous issue we tried to re-explore Istio documentation but none of the recommendations work in Kubeflow. 
This is what we currently have that works for ingress gateway:

```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: authn-filter
spec:
  workloadSelector:
    labels:
      istio: ingressgateway
  configPatches:
  - applyTo: HTTP_FILTER
    match:
      context: GATEWAY
      listener:
        filterChain:
          filter:
            name: "envoy.http_connection_manager"
            subFilter:
              name: "envoy.router"
    patch:
      operation: INSERT_BEFORE
      value:
        name: envoy.ext_authz
        config:
          http_service:
            server_uri:
              uri: http://authservice.istio-system.svc.cluster.local
              cluster: outbound|8080||authservice.istio-system.svc.cluster.local
              failure_mode_allow: false
              timeout: 10s
            authorization_request:
              allowed_headers:
                patterns:
                - exact: "cookie"
                - exact: "X-Auth-Token"
            authorization_response:
              allowed_upstream_headers:
                patterns:
                - exact: "kubeflow-userid"
          status_on_error:
            code: GatewayTimeout
```

We want to move to something like this for sidecar inbound with sidecars enabled on our KF services but nothing works. I tried the below config with sidecars enabled on KF UI services but get no redirect to our auth service. If I omit the workloadSelector and keep metadata as above I get a 504. Is there any working solution to move this filter out of ingressgateway. I am using Kubeflow v1.1.

```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: authn-filter
  namespace: kubeflow # scope to KF ns instead of entire ingress
spec:
  workloadSelector:
    labels:
      app: centraldashboard #may be changed to other UI services, this is just to test
  configPatches:
  - applyTo: HTTP_FILTER
    match:
      context: SIDECAR_INBOUND # scope inbound auth redirect to sidecars
      listener:
        filterChain:
          filter:
            name: "envoy.http_connection_manager"
            subFilter:
              name: "envoy.router"
    patch:
      operation: INSERT_BEFORE
      value:
        name: envoy.ext_authz
        config:
          http_service:
            server_uri:
              uri: http://authservice.istio-system.svc.cluster.local
              cluster: outbound|8080||authservice.istio-sytem.svc.cluster.local
              failure_mode_allow: false
              timeout: 10s
            authorization_request:
              allowed_headers:
                patterns:
                - exact: "cookie"
                - exact: "X-Auth-Token"
            authorization_response:
              allowed_upstream_headers:
                patterns:
                - exact: "kubeflow-userid"
          status_on_error:
            code: GatewayTimeout
```

**Comment by DavidSpek** [comment]
@aaron-arellano Not sure if this is related, but I think you might find it interesting and it is a new feature for Istio 1.9. https://istio.io/latest/blog/2021/better-external-authz/

**Comment by aaron-arellano** [comment]
We are not on Istio 1.9 atm, thanks for the suggestion but we want to maintain envoy filter (if possible) since this is solution community uses. I think something else should be published that shows how to enable kubeflow auth on services via sidecar.

**Comment by DavidSpek** [comment]
@aaron-arellano Kubeflow 1.3 which will be released in the coming week used Istio 1.9, so this will be what the community uses.

**Comment by aaron-arellano** [comment]
Our scenario has other services on our cluster and other teams. We don't want to hog auth for the entire ingress, just our kubeflow services. Again, I think it would be really helpful for the community to support authz on a sidecar instead of the entire ingress gateway or at least update documentation on how to achieve this. @DavidSpek, the document you linked still has the `Authorization Policy` on the ingress gateway, this is not a solution for sidecar. Again, we don't want Kubeflow resources in istio-system ns.

**Comment by aaron-arellano** [comment]
I was able to figure out a solution that works for our scenario. I use an envoy filter similar to the one above with `context: SIDECAR_INBOUND` set. I enabled sidecars on kubeflow services that use the `kubeflow-userid` so they can be aware of the cookie from the envoy filter. One thing to note, we use a third party oidc provider that uses cookie-based auth, this is why we define a cookie in our envoy filter. 

Another thing we do in our scenario is scope our kubeflow-gateway virtual service routing to a specific FQDN so the services are aware of the cookie as the FQDN we define is common to kubeflow-ui services. This FQDN is also the host on our kubeflow-gateway config yaml.

I am closing this issue as I figured out this issue on my own, hopefully my notes will help someone seeking a similar solution.
